OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,980 @@
+			 *     note that: pathDistanceSquareFactor = resourceMapIndicesColumnCount * resourceMapIndicesColumnCount + resourceMapIndicesRowCount * resourceMapIndicesRowCount,
+			 *
+			 *     Consider a map, we divide it at the length of indiceSideLength = r, and then its resourceMapIndicesColumnCount = a, resourceMapIndicesRowCount = b,
+			 *     so the map.width ≈ a*r, map.height ≈ b*r,
+			 *     the maximum euclid distance-square between two points on the map is (a*r)(a*r) + (b*r)(b*r),
+			 *     so the maximum "weight of candidate's distance to current MCV" is from 0 to -((a*r)(a*r) + (b*r)(b*r)) / (a*a + b*b) = -r*r = -indiceSideLengthSquare.
+			 *
+			 *     b) if Mobile: range depends on pathfinding distance in cell.
+			 *
+			 *     It is calculated as "pathfindDistance * pathfindDistance / pathDistanceSquareFactor".
+			 *
+			 * 2). the weight of friendly construction yard within range: -indiceSideLengthSquare. If it belongs to an ally, -indiceSideLengthSquare/2.
+			 *
+			 * 3). the weight of enemy building within range: -indiceSideLengthSquare*4.
+			 *
+			 * 4). the weight of friendly refinery within range (not for CheckBase mode): -indiceSideLengthSquare. If it belongs to an ally, -indiceSideLengthSquare/2.
+			 *
+			 * 5). the weight of resource amount (only for CheckResource mode): from 0 to +indiceSideLengthSquare/8.
+			 *
+			 *     The reason why:
+			 *
+			 *     The maximum resource amount in a indice of resource map is approximately indiceSideLengthSquare (full of it), but a stride full of resources is less likely to
+			 *     have room for buildings. So we prefer the indice have half of resource cells the most, which may give us enough room to place buildings.
+			 *
+			 *     so the weight can be: (indiceSideLengthSquare/2) - |(indiceResourceCellCount - (indiceSideLengthSquare/2))|, range from (0 to +indiceSideLengthSquare/2).
+			 *
+			 *     Note: In practive resource weight is not very important, we cannot let MCV go a long way just for a rich resource spot.
+			 *     We have to take only 1/4 of it, wich is (0 to +indiceSideLengthSquare/8),
+			 *     and apply some additional method to filter the indice for acceptable resource (not too low).
+			 */
+			var indiceSideLengthSquare = indiceSideLength * indiceSideLength;
+			switch (mcvExpansionMode)
+			{
+				/*
+				 * CheckBase mode only considers the distance to current MCV, ally construction yard within range and enemy buildings within range.
+				 * Attaction has a base value of indiceSideLengthSquare >> 3 (1/8 of the maximum distance weight, 1/(2*sqrt(2))≈ 1/2.8 of the maximum distance in map)
+				 */
+				case BotMcvExpansionMode.CheckBase:
+					var cb_conyardlocs = world.ActorsHavingTrait<Building>().Where(a => a.Owner.IsAlliedWith(player)
+						&& Info.ConstructionYardTypes.Contains(a.Info.Name)).Select(a => (a.Location, a.Owner == player)).ToArray();
```

Comment:
```suggestion
					var cb_conyardlocs = world.ActorsHavingTrait<Building>()
						.Where(a => a.Owner.IsAlliedWith(player) && Info.ConstructionYardTypes.Contains(a.Info.Name))
						.Select(a => (a.Location, a.Owner == player))
						.ToArray();
```


more readable chaining

Reply:
it does reveal that `a.Owner == player` is not `isAlly` but rather am I the owner check

Reply:
fixed